src/charm.py

```diff
@@ -3,258 +3,112 @@
-            self.unit.status = MaintenanceStatus(f"Starting {container.name} container")
-            container.pebble.replan_services()
-        else:
-            logger.debug("Pebble config unchanged")
-
-        self.unit.status = ActiveStatus()
-
-    def _is_valid_config(self) -> tuple[bool, str]:
-        """Validate required configuration.
-
-        When not configuring the agent through relations (as indicated by relation_configured
-        stored state), 'jenkins_url', 'jenkins_agent_name' and 'jenkins_agent_token' are required.
-
-        Returns:
-            Whether the configuration is valid, including a reason if it is not.
-        """
-        # Check for agent tokens
-        # mypy misinterprets this line, reports function overloading
-        if self._stored.relation_configured:  # type: ignore
-            return True, ""
-
-        # Retrieve required and non-empty configuration options
-        required_options = {"jenkins_url", "jenkins_agent_name", "jenkins_agent_token"}
-        non_empty_options = {option for option in required_options if self.model.config[option]}
-
-        if required_options.issubset(non_empty_options):
-            return True, ""
-
-        return (
-            False,
-            "Missing required configuration: "
-            f"{' '.join(sorted(required_options - non_empty_options))}",
-        )
-
-    def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
-        """Set relation data for the unit once an agent has connected.
-
-        Args:
-            event: information about the relation joined event.
-        """
-        logger.info("Jenkins relation joined")
-        num_executors = os.cpu_count()
-        config_labels = self.model.config.get("jenkins_agent_labels")
-
-        if config_labels:
-            labels = config_labels
-        else:
-            labels = os.uname().machine
-
-        event.relation.data[self.model.unit]["executors"] = str(num_executors)
-        event.relation.data[self.model.unit]["labels"] = labels
-        event.relation.data[self.model.unit]["slavehost"] = self._stored.relation_agent_name
-
-    def _on_agent_relation_changed(self, event: RelationChangedEvent) -> None:
-        """Populate local configuration with data from relation.
-
-        Args:
-            event: information about the relation changed event.
-        """
-        logger.info("Jenkins relation changed")
-
-        # Check event data
-        try:
-            # Right side data type is Union[Unit, Application]
-            relation_jenkins_url = event.relation.data[event.unit]["url"]  # type: ignore
-        except KeyError:
-            logger.warning(
-                "Expected 'url' key for %s unit in relation data. Skipping setup for now.",
-                event.unit,
-            )
-            self.model.unit.status = ActiveStatus()
+        if not self.state.jenkins_config:
             return
```

Comment:
Added methods to fallback to any relation that's existing. However, the slave relation is not backwards relatable but agent relation is. But I've left it as BlockedStatus and a comment on the code that we should implement the backwards fallback.
I think this would be a good separate ticket to limit the scope of this PR :D